# java_7

**Théorie**

Une exception est une condition que le système ou l’application rencontre et qui nécessite une sorte d’intervention ou d’action spéciale. En Java, les exceptions sont principalement utilisées pour gérer les erreurs et les autres situations exceptionnelles. Dans certains cas, vous pouvez souhaiter créer vos propres exceptions, appelées exceptions personnalisées.

Pourquoi créer des exceptions personnalisées ?

1. Clarification : Les exceptions personnalisées permettent de rendre le code plus clair et plus explicite, en fournissant des informations détaillées sur le type d’erreur qui s’est produite.
2. Précision : Elles permettent une gestion plus précise des erreurs en vous permettant de distinguer les problèmes qui peuvent survenir dans votre code.
3. Flexibilité : Avec les exceptions personnalisées, vous pouvez définir des comportements d’exception spécifiques à vos besoins.

Comment créer des exceptions personnalisées ?

Pour cela, vous devrez créer une nouvelle classe qui hérite de la classe `Exception` de Java ou l’une de ses sous-classes. Vous pouvez alors ajouter des champs et des méthodes supplémentaires si vous le souhaitez.

```java
public class MyException extends Exception {    public MyException(String message) {        super(message);    }}
```

**Exercice pratique**

Voici un simple programme Java qui utilise une exception personnalisée.

```java
// Notre exception personnalisée
class InvalidAgeException extends Exception {
    public InvalidAgeException(String message) {
        super(message);
    }
}

class Main {
    // Méthode qui vérifie l'âge
    static void checkAge(int age) throws InvalidAgeException {
        if (age < 18) {
            throw new InvalidAgeException("Not Eligible for voting");
        } else {
            System.out.println("Eligible for voting");
        }
    }

    public static void main(String args[]) {
        try {
            checkAge(13);
        } catch (InvalidAgeException e) {
            System.out.println(e.getMessage());
        }
    }
}
```

Dans cet exemple, nous avons créé une exception personnalisée `InvalidAgeException`. Notre méthode `checkAge` vérifie l’âge et lance une `InvalidAgeException` si l’âge est inférieur à 18. Dans notre méthode `main`, nous appelons `checkAge` dans un bloc try/catch et affichons le message de l’exception si une `InvalidAgeException` est lancée.